substrate/primitives/version/src/lib.rs

```diff
@@ -230,9 +221,406 @@ pub struct RuntimeVersion {
+				let mut impl_name: Option<RuntimeString> = None;
+				let mut authoring_version: Option<u32> = None;
+				let mut spec_version: Option<u32> = None;
+				let mut impl_version: Option<u32> = None;
+				let mut apis: Option<ApisVec> = None;
+				let mut transaction_version: Option<u32> = None;
+				let mut system_version: Option<u8> = None;
+
+				while let Some(key) = map.next_key()? {
+					match key {
+						Field::SpecName => {
+							if spec_name.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"specName",
+								));
+							}
+							spec_name = Some(map.next_value()?);
+						},
+						Field::ImplName => {
+							if impl_name.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"implName",
+								));
+							}
+							impl_name = Some(map.next_value()?);
+						},
+						Field::AuthoringVersion => {
+							if authoring_version.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"authoringVersion",
+								));
+							}
+							authoring_version = Some(map.next_value()?);
+						},
+						Field::SpecVersion => {
+							if spec_version.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"specVersion",
+								));
+							}
+							spec_version = Some(map.next_value()?);
+						},
+						Field::ImplVersion => {
+							if impl_version.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"implVersion",
+								));
+							}
+							impl_version = Some(map.next_value()?);
+						},
+						Field::Apis => {
+							if apis.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field("apis"));
+							}
+							apis = Some({
+								struct DeserializeWith<'de> {
+									value: ApisVec,
+									lifetime: PhantomData<&'de ()>,
+								}
+								impl<'de> serde::Deserialize<'de> for DeserializeWith<'de> {
+									fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+									where
+										D: serde::Deserializer<'de>,
+									{
+										Ok(DeserializeWith {
+											value: apis_serialize::deserialize(deserializer)?,
+											lifetime: PhantomData,
+										})
+									}
+								}
+
+								map.next_value::<DeserializeWith<'de>>()?.value
+							});
+						},
+						Field::TransactionVersion => {
+							if transaction_version.is_some() {
+								return Err(<A::Error as serde::de::Error>::duplicate_field(
+									"transactionVersion",
+								));
+							}
+							transaction_version = Some(map.next_value()?);
+						},
+						Field::SystemVersion =>
+							if let Some(system_version) = system_version {
+								let new_value = map.next_value::<u8>()?;
+								if system_version != new_value {
+									return Err(<A::Error as serde::de::Error>::custom(
+										alloc::format!(
+											r#"Duplicated "stateVersion" and "systemVersion" \
```

Comment:
You don't really know this. It could also be that there are more than one `stateVersion`.

Reply:
You're right, but it is one of those two fields. And I decided to make this assumption in an error message since it is the most likely. Open for suggestions.